File: 1. Caesar Cipher/CaeserCipherCrackingLanguagesDetecting.py
# When Cracking a given cipher it nay be useful to detect whether
# the decrypted language is eglish or not
ALPHABET=" ABCDEFGHIJKLMNOPQRSTUVWXYZ"

ENGLISH_WORDS = []

def get_data():
    # opening english words
    words = open("./english_words.txt",'r')
    # we can use a dictionary and check whether the given words are
    # are present in a dictionary or not
    
    for word in words.read().split("\n"):
        ENGLISH_WORDS.append(word)
    
    words.close()
    print(len(ENGLISH_WORDS))

#count the number of english words in a given text
def count_words(text):
    text = text.upper()
    # transform the text into a list of words (split by spaces)
    words = text.split(' ')
    # matches counts the number of english words in the text
    matches = 1
    # consider all the words in the text and check whether the
    # given word in english or not
    for word in words:
        if word in ENGLISH_WORDS:
            matches +=1
    
    return matches

#decides whether a given text is english or not
def is_text_english(text):
    # number of English words in a given text
    matches = count_words(text)
    # you can define your own classification alogrithm in this case
    # the assumption: if 70% of the words in the text are english words
    # then it is an enlgish text
    if (float(matches) / len(text.split(' ')))* 100 >= 70:
        return True
    
    #not an english text
    return False

# now we will try brute force to get the decrepted text
def crake_caesar(cipher_text):
   # we try all the possible key values so the size of the ALPHABET
    for key in range(len(ALPHABET)):

        # reinitialize this to be an empty string
        plain_text = ''

        # we just have to make a simple caesar decryption
        for c in cipher_text:
            index = ALPHABET.find(c)
            index = (index - key) % len(ALPHABET)
            plain_text = plain_text + ALPHABET[index]

        # print the actual decrypted string with the given key
        if is_text_english(plain_text):
            print("We have managed to crack Caesar cipher, the key is: %s, the message is %s" % (key, plain_text))


if __name__ =="__main__":
    get_data()
    # plain_text = "The most important property of a program is that it is correct. For the Caesar Cipher, Latin letters are rotated, in either direction, by a fixed shift amount. Decoding is the reverse of encoding."
    # encrypted = "KHOOR"
    encrypted = "FU@SWRJUDSK@CDQGCKDVKLQJAB"
    print(crake_caesar(encrypted))
    
    
    

File: 1. Caesar Cipher/CaeserCipherCrackingLanguagesDetecting.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"math"
	"os"
	"strings"
)

var ENGLISH_WORDS []string

const TEXT = " ABCDEFGHIJKLMNOPQRSTUVWXYZ!@#$%^&*(){<>?}0987654321"

func main() {
	// text := "The most important property of a program is that it is correct. For the Caesar Cipher, Latin letters are rotated, in either direction, by a fixed shift amount. Decoding is the reverse of encoding."
	get_data()

	text := "FU@SWRJUDSK@CDQGCKDVKLQJAB"
	crake_caesar(text)
}
func get_data() {
	file, err := os.Open("./english_words.txt")
	// we can use a dictionary and check whether the given words are
	// are present in a dictionary or not

	if err != nil {
		log.Fatal("failed to open")
	}

	scanner := bufio.NewScanner(file)

	scanner.Split(bufio.ScanLines)

	for scanner.Scan() {
		ENGLISH_WORDS = append(ENGLISH_WORDS, scanner.Text())
	}

	file.Close()
	fmt.Println(len(ENGLISH_WORDS))

}

func count_words(text string) int {
	text = strings.ToUpper(text)
	// transform the text into a list of words (split by spaces)
	mystr := strings.Split(text, ` `)

	str := make(map[string]bool)
	for _, w := range mystr {
		str[string(w)] = true
	}
	// counts the number of english words in the text
	var count int
	count = 1

	// consider all the words in the text and check whether the
	// given word in english or not

	for _, w := range mystr {
		if str[w] {
			count += 1
		}
	}

	return count

}

func is_text_english(text string) bool {
	// number of English words in a given text
	matches := count_words(text)
	// you can define your own classification alogrithm in this case
	// the assumption: if 50% of the words in the text are english words
	// then it is an enlgish text
	if (float64(matches)/float64(len(strings.Split(text, ` `))))*100 >= 50 {
		return true
	} else {
		return false
	}
}

func crake_caesar(str string) {
	// we try all possible key value
	for key := 0; key < len(TEXT); key++ {
		decode_text := ""
		for _, char := range str {
			idx := strings.Index(TEXT, string(char))
			//  formula how we can decrypt the text
			//  Dn(x) = (x - n) % len(ALPHABET)
			f := (idx - key) % len(TEXT)
			if f < 0 {
				// if f is negative value
				// f = f + len(TEXT)
				f = len(TEXT) - int(math.Abs(float64(f)))
			}
			decode_text += string(TEXT[f])

		}
		if is_text_english(decode_text) {
			fmt.Printf("With key %v the result is %v \n", key, decode_text)
		}
	}
}
